### PR TITLE
[Android] Placepage doesn't exit fullscreen mode

### PR DESCRIPTION
**Pull request by strump**
As @pastk asked in [comment](https://github.com/organicmaps/organicmaps/pull/8443#issuecomment-2185319015)

Changed android behaviour: in fullscreen mode opening a placepage doesn't exit fullscreen mode.

https://github.com/organicmaps/organicmaps/assets/720808/9ad2d727-ff52-42c8-89e2-0a04010a6b7c

## Update 2024-06-25

Now OM exits fullscreen mode before starting navigation.

Also in route planning mode you can enter/exit fullscreen mode with a long tap.

### CODE REVIEW

**Comment by pastk** [comment]
Thanks for a quick fix!

Now UI doesn't reappear on short taps - good.

But there are other bugs I've noticed.

1. UI doesn't re-appear after entering routing mode. Either is should re-appear or be switchable by a long-tap (its not switchable).
Same for the navigation mode. @biodranik @kirylkaveryn how is it on iOS?

2. PP buttons disappear in open PP or even whole PP disappears permanently:
- open a PP
- long tap to hide UI (PP stays)
- long tap again to bring UI back (PP stays)
- tap elsewhere to have another place in PP
- PP will refresh but without PP buttons (route to...)
image
- close PP
- tap somewhere to open PP - it won't open!


3. (minor) when UI is hidden and PP is open there is a wasted space under PP which is normally used by system buttons (back, home...) - I guess depends on system UI settings, I have system buttons enabled

**Comment by strump** [comment]
@pastk Which android version do you use?

I can't reproduce issue 2 on Android 14.

**Comment by pastk** [comment]
Android 9.
It seems the issue is in internal app (PP) state. As another long tap makes PP functional again.

**Comment by kirylkaveryn** [comment]
@pastk only the 1) bug partly exists in iOS.

The "fullscreen mode" switching is disabled during the planning the route, but all controls _always visible during the navigation mode_ and cannot be disabled in ios.

**Comment by biodranik** [comment]
@strump @kirylkaveryn can you please finish the fix with what was discussed here and in https://github.com/organicmaps/organicmaps/pull/8443#issuecomment-2187020690 ?

**Comment by kirylkaveryn** [comment]
> @strump @kirylkaveryn can you please finish the fix with what was discussed here and in https://github.com/organicmaps/organicmaps/pull/8443#issuecomment-2187020690 ?

Sure! 
Let's revise the desired behaviour:
1. Long tap activates the Fullscreen mode and hide both the PP and the controls.
2. Second long tap deactivates the fullscreen mode and put back the controls.
3. Controls always visible on navigation, planning the route, searching.

**Comment by biodranik** [comment]
Yes. @pastk any other ideas or improvements?

P.S. we may find a good use for a long tap during the navigation and routing too.

**Comment by kirylkaveryn** [comment]
Additional questions:
1. If the fullscreen is enabled and the user opens the PP, the controls should be hidden?
2. f the fullscreen is enabled, the PP is opened and the user longtaps on map, should we hide the PP and show the controls? or we should only exit the fullscreen mode without hiding the PP?

**Comment by pastk** [comment]
> 2\. Second long tap deactivates the fullscreen mode and put back the controls.

I'd rather have **any** tap to deactivate the fullscreen mode.

But I'm fine if you do as you've suggested - hopefully I'm wrong and users will find their way out of FS mode easily :)

> * If the fullscreen is enabled and the user opens the PP, the controls should be hidden?

I think yes.

> 
> * f the fullscreen is enabled, the PP is opened and the user longtaps on map, should we hide the PP and show the controls? or we should only exit the fullscreen mode without hiding the PP?

I think the latter option would be more consistent.

**Comment by kirylkaveryn** [comment]
This solution 

> I'd rather have **any** tap to deactivate the fullscreen mode.

is not compatible with 

> > * If the fullscreen is enabled and the user opens the PP, the controls should be hidden?
> 
> I think yes.
> 
> >

In this case, no way to hide the controls and open the PP (because PP opening will trigger the exit from the fullscreen).

**Comment by pastk** [comment]
> 
> In this case, no way to hide the controls and open the PP (because PP opening will trigger the exit from the fullscreen).

Yes, this is exactly how I'd prefer it as I see very little use of having a PP open but controls hidden at the same time.

But if you're going the "PP is openable in the FS mode" then IMO controls should stay hidden as the app is in FS mode still, right?
If you do it like "when in the FS mode PP is opened then controls appear and after PP is closed they disappear again" - I think its more confusing / less consistent with the "FS mode" concept.

**Comment by kirylkaveryn** [comment]
This behavior

> I'd rather have **any** tap to deactivate the fullscreen mode.

is not compatible with the

> > * If the fullscreen is enabled and the user opens the PP, the controls should be hidden?
> 
> I think yes.

because the PP opening will exit the fullscreen.
I prefer to exit from the fullscreen on any tap too.

**Comment by pastk** [comment]
> I prefer to exit from the fullscreen on any tap too.

Then do it? :)

File: android/app/src/main/java/app/organicmaps/MwmActivity.java [hunk 1]
```diff
@@ -1249,9 +1248,7 @@ public void onSwitchFullScreenMode()
 
   private void setFullscreen(boolean isFullscreen)
   {
-    if (RoutingController.get().isNavigating()
-        || RoutingController.get().isBuilding()
-        || RoutingController.get().isPlanning())
```
Comment (strump) [inline android/app/src/main/java/app/organicmaps/MwmActivity.java:1254]: This 'if' block was added in 2016 in commit [e66fe5b](https://github.com/organicmaps/organicmaps/commit/e66fe5b62d63094f6781aafe9025a614303a33ba). I reduced it to only `isNavigation()` condition. Any objections?

**Comment by biodranik** [comment]
>>  I prefer to exit from the fullscreen on any tap too.

> Then do it? :)

If a user deliberately kept a finger to hide the UI, then he/she is ready to repeat it again, when UI is needed, right?

Showing the UI back on a tap to see some POI info requires keeping a finger to hide the UI again and again. This is annoying.

**Comment by pastk** [comment]
> If a user deliberately kept a finger to hide the UI, then he/she is ready to repeat it again, when UI is needed, right?

Not necessarily, it could be accidental. E.g. my mom often does a long tap instead of a short tap and just can't change that habit :/

**Comment by biodranik** [comment]
Long taps may be implemented and work differently in different OSes/apps. Did your mom do accidental long taps in OM?

For accidental cases (which hopefully happens rarely), long-tapping again to exit full-screen mode is not as annoying as always long-tapping again to enable full-screen after each PP opening.

Note that there is a toast that hints at how to exit the full-screen mode for such accidental clicks.

**Comment by biodranik** [comment]
> Leaving FS mode by any tap is the release behavior that been there for many years - no complaints :)

Do not compare apples with motorcycles. A tap on an empty space was used as a switch to enable/disable full-screen mode. Selecting a POI _did not toggle the full-screen mode_. Now a long tap is used to enable/disable full-screen mode, not a single tap/PP opening.

**Comment by pastk** [comment]
> A tap on an empty space was used as a switch to enable/disable full-screen mode.

An empty space tap entered FS mode. And then **any tap anywhere** disabled the FS mode. Thats how its been working on Android releases for years.

**Comment by biodranik** [comment]
I don't have an Android device in my hands now to check it. But the correct behavior I'm describing has been on iOS since the beginning for more than a decade (the iOS version was always developed first).

**Comment by pastk** [comment]
You mean to leave the FS mode on iOS the only way is to short tap on an empty space?

I.e. if in FS mode one taps on a POI then it won't be selected? And to do it one has to tap on an empty space first to disable FS mode?

**Comment by biodranik** [comment]
> You mean to leave the FS mode on iOS the only way is to short tap on an empty space?

Yes.

> I.e. if in FS mode one taps on a POI then it won't be selected? And to do it one has to tap on an empty space first to disable FS mode?

No. Tap on a POI shows PP as expected. But the interface buttons do not come back.

Found an Android device. Indeed the behavior is different in the production. And it is annoying! I use full screen to see more of the map, and then to check info about objects, or current speed/elevation. Turning it always back on with a long tap after selecting any POI is a lot of overhead (many long taps).

**Comment by biodranik** [comment]
This appeared in API 34 emulator after a long tap. What is it? :)

выява

**Review by biodranik — COMMENTED** [review]
@strump thanks! It works, the only thing to fix is to also hide the Place Page when the interface and PP are visible, and a long tap is done.

**Review by biodranik — APPROVED** [review]
This can be merged, and the follow-up like-on-ios behavior change can be done later. @vng PTAL

**Comment by pastk** [comment]
> I can't reproduce issue 2 on Android 14.

Is issue 2 reproduced before this PR?

**Review by vng — APPROVED** [review]
LGTM, please, double check on devices with different use cases.

**Comment by strump** [comment]
> This can be merged, and the follow-up like-on-ios behavior change can be done later. @vng PTAL

Please don't merge for now. I'm in progress of fixing bottom buttons issue on Android 9. Will notify when it's ready

**Comment by biodranik** [comment]
@strump any chance to get this fix into the release?

CC @rtsisyk

**Comment by strump** [comment]
Problem with Android 9 reported by @pastk and reproducible in 100% cases:

1. Open place page
2. Use long tap to enter fullscreen mode
3. Keeping place page open exit fullscreen mode with another long tap
4. Try to open any POI to refresh place page content

**Actual result:** Place page layout is not updated.

On POI tap Organic Maps refreshes place page content: updates title, address, details, refreshes bottom buttons. But after exit from fullscreen layout is not refreshed at all. Organic Maps replaces bottom buttons but they have 0px height and 0px width. Labels' text is changed but layout is not performed in `place_page_container_fragment`:

place_page_layout

Inside `PlacePageView.onViewCreated(...)` method we add important layout listener:

```java
mFrame.addOnLayoutChangeListener((v, left, top, right, bottom, oldLeft, oldTop, oldRight, oldBottom) -> {
  // Listener body
});
```
But the listener body is **never called after OM exits fullscreen** mode. I don't know why layout is not refreshed.

The problem is caused by `UiUtil.setFullscreen(...)` function:
```java
wic.hide(WindowInsetsCompat.Type.systemBars());
```
Commenting out this line makes layout work.

### Workaround:

I found a workaround: for Android API < 30 use deprecated method:
```java
window.setFlags(FLAG_FULLSCREEN, FLAG_FULLSCREEN);
```

For Android API 30+ I left existing fullscreen enter/exit code.

The problem with this approach is that bottom button are not hidden in Android 9:

android-9-bottom-bar

Also need to test on all affected Android versions.

File: android/app/src/main/java/app/organicmaps/MwmActivity.java (hunk 1 above)
Comment (biodranik) [reply android/app/src/main/java/app/organicmaps/MwmActivity.java:1254]: @strump does it mean that the interface will hide when the route is planned or is built? That should not happen. Full screen is disabled on iOS in such cases.

**Comment by pastk** [comment]
I've tested on my A9 and the bug is gone, thanks!

> The problem with this approach is that bottom button are not hidden in Android 9:
> 
> android-9-bottom-bar

When PP is open I think its actually better with system buttons than having a blank unused space there (like it was before the fix) :)

Yeap, when there is no PP system buttons are present still so its not 100% fullscreen, but IMHO its OK (if hard to find a better fix).
Actually allowing to open a PP in a "fullscreen" mode is much less "fullscreen" than having those buttons :)

**Review by pastk — APPROVED** [review]

**Comment by biodranik** [comment]
A proper fix should not show bottom system buttons, as it was before this PR, and also reduce the height of the Place Page so its buttons go down and there would be no empty space (the height is reduced on the Android 12 emulator and likely other 11+ versions with this PR).

Note that @arnaudvergnet had many issues with full screen refactoring and insets, maybe if you read comments for related PRs, something relevant can be found: https://github.com/organicmaps/organicmaps/pulls?q=is%3Apr+is%3Aclosed+insets

**Comment by biodranik** [comment]
This workaround can be acceptable for the upcoming release after testing on Android 11, 13, 14, 15. Looks like there is a bug in the current implementation when layout/fragment/some views become invalid after switching. Maybe checking all `setOnApplyWindowInsetsListener` calls can help.

**Comment by strump** [comment]
@biodranik I did a quick check: found all `.setOnApplyWindowInsetsListener` and commented out. But the issue with broken layout is still present in Android 9. I'll look closer on window insets but have doubts that those are the reason of the bug.

**Comment by biodranik** [comment]
Okay, looks like `addOnLayoutChangeListener` is called every second on location update, to redraw the direction arrow and distance. The question is "Why does it cause the total relayout of the whole place page"? That is unlikely related, a separate performance/battery issue which can be optimized later.

**Comment by strump** [comment]
@biodranik  blue arrow with distance to a POI is updated every time we get GPS coordinates. Label with distance is changed which leads to layout update. Disabling GPS can stop those updates.

**Comment by biodranik** [comment]
@strump looks like commenting this code helps:
```
//    mFrame.addOnLayoutChangeListener((v, left, top, right, bottom, oldLeft, oldTop, oldRight, oldBottom) -> {
//      final int oldHeight = oldBottom - oldTop;
//      final int newHeight = bottom - top;
//      if (oldHeight != newHeight)
//        mPlacePageViewListener.onPlacePageContentChanged(mPreview.getHeight(), newHeight);
//    });
```

onPlacePageContentChanged is called from another place. There is one scenario where it is not yet properly called after commenting. Looks like calling `onPlacePageContentChanged` causes issue on older devices (it reproduces starting from Android 6, I didn't check Android 5.

**Comment by strump** [comment]
> @strump looks like commenting this code helps:
> 
> ```
> //    mFrame.addOnLayoutChangeListener((v, left, top, right, bottom, oldLeft, oldTop, oldRight, oldBottom) -> {
> //      final int oldHeight = oldBottom - oldTop;
> //      final int newHeight = bottom - top;
> //      if (oldHeight != newHeight)
> //        mPlacePageViewListener.onPlacePageContentChanged(mPreview.getHeight(), newHeight);
> //    });
> ```
> 
> onPlacePageContentChanged is called from another place. There is one scenario where it is not yet properly called after commenting. Looks like calling `onPlacePageContentChanged` causes issue on older devices (it reproduces starting from Android 6, I didn't check Android 5.

Proposed patch doesn't work for Android 9. Layout isuue is not resolved.

**Comment by pastk** [comment]
Merge it and take into the release?
Its much better than the currently buggy master.

File: android/app/src/main/java/app/organicmaps/MwmActivity.java (hunk 1 above)
Comment (strump) [reply android/app/src/main/java/app/organicmaps/MwmActivity.java]: Reverted this change. Now Android exits fullscreen to start route planning. In route planning mode no fullscreen mode allowed.

**Review by biodranik — APPROVED** [review]
Thanks! Now the behavior is equal to iOS: PP is hidden when FS mode is activated.

**Comment by biodranik** [comment]
@rtsisyk please add this commit into the release: https://github.com/organicmaps/organicmaps/commit/b522549c79c3093b1235b6a8c1b783afad9709bf

**Pull request merged**